File: python/leetcode/061.py
class Solution(object):
    def rotateRight(self, head, k):
        if not head or not head.next or k == 0:
            return head
        n, slow = 0, head
        while slow:
            n += 1
            slow = slow.next
        if k % n == 0:
            return head
        k %= n
        slow, fast = head, head
        for i in range(k):
            fast = fast.next
        while fast.next:
            slow = slow.next
            fast = fast.next
        fast.next = head
        head = slow.next
        slow.next = None
        return head

File: cpp/leetcode/061.cpp
class Solution {
public:
    ListNode *rotateRight(ListNode *head, int k) {
        if (!head || k == 0) {
            return head;
        }
        int cnts = 0;
        for (ListNode *fast = head; fast; fast = fast->next) {
            cnts++;
        }
        k %= cnts;
        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        ListNode *fast = dummy, *slow = dummy;
        for (int i = 0; i < k; i++) {
            fast = fast->next;
        }
        for (; fast->next; fast = fast->next, slow = slow->next) ;
        fast->next = head;
        head = slow->next;
        slow->next = NULL;
        return head;
    }
};